contracts/03-tamagotchi-nft/io/src/lib.rs

```diff
@@ -1,40 +1,135 @@
 #![no_std]
 
-use gmeta::Metadata;
-use gstd::prelude::*;
+use gmeta::{In, InOut, Metadata, Out};
+use gstd::{exec, prelude::*, ActorId};
+
+pub const HUNGER_PER_BLOCK: u64 = 1;
+pub const ENERGY_PER_BLOCK: u64 = 2;
+pub const BOREDOM_PER_BLOCK: u64 = 2;
+pub const FILL_PER_SLEEP: u64 = 1000;
+pub const FILL_PER_FEED: u64 = 1000;
+pub const FILL_PER_ENTERTAINMENT: u64 = 1000;
 
 #[derive(Default, Encode, Decode, TypeInfo)]
 #[codec(crate = gstd::codec)]
 #[scale_info(crate = gstd::scale_info)]
 pub struct Tamagotchi {
     // TODO: 0️⃣ Copy fields from previous lesson and push changes to the master branch
+    pub name: String,
+    pub date_of_birth: u64,
+    pub owner: ActorId,
+    pub fed: u64,
+    pub fed_block: u64,
+    pub entertained: u64,
+    pub entertained_block: u64,
+    pub rested: u64,
+    pub rested_block: u64,
     // TODO: 1️⃣ Add new fields
+    pub approved_account: Option<ActorId>,
+}
+
+impl Tamagotchi {
+    pub fn sleep(&mut self) {
+        let blocks_height = blocks_height();
+        let updated_rested = updated_field_value(
+            self.rested,
+            self.rested_block,
+            ENERGY_PER_BLOCK,
+            blocks_height,
+        );
+        self.rested = update_field(updated_rested, FILL_PER_SLEEP);
+        self.rested_block = blocks_height;
+    }
+
+    pub fn feed(&mut self) {
+        let blocks_height = blocks_height();
+        let updated_feed =
+            updated_field_value(self.fed, self.fed_block, HUNGER_PER_BLOCK, blocks_height);
+        self.fed = update_field(updated_feed, FILL_PER_FEED);
```

Comment:
I guess ` updated_field_value` and `update_field` might be combined into one function.